_reference/05_the_view_macro.adoc

```diff
@@ -0,0 +1,234 @@
+                gtk::Button {
+                    label: "-",
+                },
+            },
+        }
+    }
+----
+
+In this setup, we declared a gtk window, containing a gtk box, containing two gtk buttons.
+We also set the `label` property of each button.
+Everytime we nest a widget within the `{}` of another widget, we set up a hierarchical parent/child or container/child relationship.
+
+The widgets contained in the `view` can be gtk widgets as well as relm widgets.
+
+=== Relm widgets
+
+Here is an example of adding a relm widget in a view hierarchy:
+
+
+[source,rust]
+----
+    SearchView((None, SearchItemsType::All, OperationMode::ItemActions)) {
+    }
+----
+
+In this case, `SearchView` is the name of the relm component, and in the brackets we have all the component's model parameters.
+
+As a reminder, when we declare a relm component, we must define its `fn model()` function. That function takes as as second parameter the model parameters.
+For instance, in this case:
+
+[source,rust]
+----
+    fn model(
+        relm: &relm::Relm<Self>,
+        params: (
+            Option<String>,
+            SearchItemsType,
+            OperationMode,
+        ),
+    ) -> Model {
+----
+
+The component model parameter has three parameters, which we pass through a tuple when building the `view!` hierarchy.
+
+=== Widget attributes
+
+We can also attach attributes to widgets. For instance:
+
+[source,rust]
+----
+    view! {
+        #[name="toplevel"]
+        #[style_class="mainwin"]
+        gtk::Window {
+        }
+    }
+----
+
+Let's now review the possible widget attributes.
+
+- <<The 'name' widget attribute>>,
+- <<The 'style_class' widget attribute>>,
+
+==== The 'name' widget attribute
+The `name` widget attribute:
+
+[source,rust]
+----
+#[name="toplevel"]
+----
+
+Allows you to access the widget easily in your relm code.
+When you're in the `Widget` implementation, you can then access the widget through `self.toplevel`, or the name you've given.
+This is often useful to initialize the widget in your `init_view()` method, although generally speaking it's better if you can manage to do everything within the `view!` macro.
+It's not always possible though.
+
+==== The 'style_class' widget attribute
+
+[source,rust]
+----
+#[style_class="mainwin"]
+----
+
+Allows to add a gtk style class name to the widget. This is equivalent to running this code in your `init_view()`:
+
+[source,rust]
+----
+widget.get_style_context().add_class("mainwin");
+----
+
+Because it's possible to add multiple CSS classes to a single widget, you can add multiple `style_class` annotations to a single widget.
+In that case all the classes that you list will be added to the widget.
+
+=== Gtk child properties
+
+When declaring the widget hierarchy, we will sometimes have to specify link:https://developer.gnome.org/gtk3/stable/GtkContainer.html#id-1.3.20.3.10.8[gtk child properties].
+
+This can be done through a `child` subnode in the view hierarchy:
```

Comment:
I'm correcting this, but I also since learned that child properties don't exist anymore in gtk4. Still, probably makes sense to mention those, hard to achieve much at all in gtk3 without them.

Reply:
sorry, I don't completely understand this. In relm master examples, I can see:

```rust
            MyFrame {
                child: {
                    padding: 10,
                },
            }
```
And it seems to work, but I can't find a gtk function `set_child_padding` anywhere?

Reply:
The method is [there](http://gtk-rs.org/docs/gtk/trait.BoxExt.html#tymethod.set_child_packing).
The method is on `Box` and not on the `MyFrame` because child properties are called on the parent.

Reply:
[It's hidden.](https://github.com/gtk-rs/gtk-rs/blob/8a4c2d25e2cd40a6399e919908cde97ea0e93c1f/gtk/src/auto/box_.rs#L487) If I recall correctly, those were added for `relm` specifically.

Reply:
it seems like it can get very technical very quickly. reading again my original text, it seems technically correct, if incomplete. Although it's true it's supposed to a reference. I'm saying "you'll probably need to set up gtk child properties, and you can use this mechanism". This seems correct. It's true that this mechanism can be used for other purposes too, but even though this is a reference, maybe we should list some expected uses rather than going in the nitty-gritty of hidden functions?

Reply:
The current text is enough for now.